### ISSUE TITLE
Move CourtListener to Co-Location Host

### ISSUE THREAD

**Issue by mlissner** [freelawproject/courtlistener#1082]
I've been working on this for about six months, and I'm happy to share that I'm on the brink of signing a co-location agreement with [LMI.net](https://lmi.net), a local provider here in Berkeley. Their prices are fair, their staff is smart, and their location is near. 

Before doing this, we need to make a detailed plan of how the move will be completed. Here's what comes to my mind, but if anybody reading this has experience with this kind of thing, I'd welcome suggestions. 

Here's what's on my mind so far:

 - [ ] DNS.

    I imagine the DNS part of this is pretty simple. A few days before the move, tweak our TTLs down to, say 30s or something, and then on the day of, tweak the DNS configuration to point to our new IP address.

 - [ ] Reverse DNS

    I don't think this is actually set up by our host at present, so I think we just need to get LMI to set this up before the move. As far as I know, this can be done at literally any time.

 - [ ] Replication

    We're replicating from our mega server to a server in AWS and from there to a number of clients. When we move our servers, we're going to sever the connection between our server and AWS, and we're going to change the IP address AWS uses to connect to us. 

    I just checked, and the subscription in AWS uses our hostname, so it shouldn't need an update, but I don't know how often it refreshes its DNS cache or how to forcibly flush it. 

I actually can't think of much else. DNS should take care of most stuff for us. Can somebody please tell me the 50 things I'm forgetting, because I feel foreboding on this one.

**Comment by mlissner** [comment]
Here's another one. 

 - [ ] We need to update our AWS security groups to allow ingress/egree to/from our new IP address. Here's the link for once we get the address: https://us-west-2.console.aws.amazon.com/ec2/home?region=us-west-2#SecurityGroups:search=sg-044759e3ba310566a;sort=groupId

**Comment by mlissner** [comment]
Reverse DNS is set for our new IP:

```
12:59:42::mlissner@new-courtlistener::/home/restic/bin 
↪ dig -x 66.117.145.178

; <<>> DiG 9.9.5-3ubuntu0.19-Ubuntu <<>> -x 66.117.145.178
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 59816
;; flags: qr rd ra; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 4096
;; QUESTION SECTION:
;178.145.117.66.in-addr.arpa.	IN	PTR

;; ANSWER SECTION:
178.145.117.66.in-addr.arpa. 900 IN	PTR	courtlistener.com.

;; Query time: 11 msec
;; SERVER: 64.201.245.201#53(64.201.245.201)
;; WHEN: Tue Dec 17 12:59:47 PST 2019
;; MSG SIZE  rcvd: 87
```

**Comment by mlissner** [comment]
Using our IP, I dug up what LMI has in terms of IP addresses. Never did this before, but it's kind of neat: https://rdap.arin.net/registry/ip/66.117.128.0.  That shows that they have 8192 ip addresses allocated. I suppose if we wanted to have some real fun, we could see what else shows up on those addresses (or others could find us that way too in the future). Not sure the point, but it's interesting.

**Comment by mlissner** [comment]
Our IP address is not blacklisted: https://mxtoolbox.com/SuperTool.aspx?action=blacklist%3a66.117.145.178&run=toolpage

**Comment by mlissner** [comment]
Looks like replication won't be too bad either. I asked in the Postgres slack, and got the following response:

![Screenshot from 2019-12-17 13-28-01](https://user-images.githubusercontent.com/236970/71035777-12917e00-20d1-11ea-8615-35dc10005162.png)

I think this should be fine. The publisher sends changes to the subscriber, so I think the only thing that matters is that the subscriber knows where the publisher is. I think what we do is stop the subscription, with:

    ALTER SUBSCRIPTION mysub DISABLE;

Then cut the power to the publisher. Move it to the new building. Then start up the publisher and re-enable the subscriber, with:

    ALTER SUBSCRIPTION mysub ENABLE;

(as described here: https://pgdash.io/blog/postgres-replication-gotchas.html)

And from [the docs](https://www.postgresql.org/docs/10/sql-altersubscription.html):

![Screenshot from 2019-12-17 13-32-53](https://user-images.githubusercontent.com/236970/71036089-c561dc00-20d1-11ea-870c-aa9da3a84638.png)


If that doesn't work, we might have DNS cached somewhere, but we should be able to fix that by just ALTERing the subscription to use the new IP address instead of using the hostname (as it is now).

**Comment by mlissner** [comment]
We had two A records that pointed to our current IP. I've set them both with a one minute TTL (they both previously had five minutes). 

That should take care of DNS until we're ready to change their IP addresses to the new ones.

**Comment by mlissner** [comment]
OK, ingress and egress to AWS are established in AWS via the link above. We'll want to remove the old firewall rules once the move is complete.

**Comment by mlissner** [comment]
Another todo:

 - [ ] Add ingress/egress to/from all of our replication customer's AWS locations. It's either that or I need a server left behind at our old IP that I can use to get that IP address.

**Comment by mlissner** [comment]
This migration is planned for tomorrow evening, PST. Realtime notes are being kept here, including notes about moving the hardware: 

https://docs.google.com/document/d/11MHLzuHjCNaglrqExEV_VqmRqzuEwMLIlnxorYpXlVI/edit?usp=sharing

**Comment by mlissner** [comment]
Welp, we've postponed the move, due to the Bay Area being on lock down: https://www.sfchronicle.com/local-politics/article/Bay-Area-must-shelter-in-place-Only-15135014.php?t=ebc54dc680

I'll revisit this in a few weeks, when hopefully life will be back to normal a bit.

**Comment by mlissner** [comment]
A few things to restore after the move:

1.  sudo docker run --rm -it -v /sata:/sata -v /home/mlissner/.aws/:/root/.aws amazon/aws-cli s3 sync /sata/recap 's3://com-courtlistener-storage/recap' --profile storage --acl public-read; bush "All done uploading"

2. Speed test before reveals:

01:56:16::mlissner@new-courtlistener::~ 
↪ speedtest-cli
Retrieving speedtest.net configuration...
Retrieving speedtest.net server list...
Testing from Paxio (64.201.244.226)...
Selecting best server based on latency...
Hosted by Razzolink Inc (San Jose, CA) [1.70 km]: 4.914 ms
Testing download speed........................................
Download: 456.54 Mbit/s
Testing upload speed..................................................
Upload: 444.18 Mbit/s

3. Disable the replication:

        alter subscription replicasubscription2 disable;

**Comment by mlissner** [comment]
## After the move!

Speed test:

```
↪ speedtest-cli
Retrieving speedtest.net configuration...
Retrieving speedtest.net server list...
Testing from LanMinds (66.117.145.178)...
Selecting best server based on latency...
Hosted by Unwired (San Francisco, CA) [18.00 km]: 5.143 ms
Testing download speed........................................
Download: 877.20 Mbit/s
Testing upload speed..................................................
Upload: 444.53 Mbit/s
```

Not bad.

2. Re-enabled replication:

         alter subscription replicasubscription2 enable;

    Verified with: https://www.courtlistener.com/api/replication/status/

3. NFS came up fine.

4. Redis came up fine.

5. Postgres needed a manual boot.

6. Solr needed to be started with `docket start solr`.

7. Celery needed to be started with the usual scale command.

8. Nginx and friends needed to be started with docker compose up. 

9. Our canary needed to be plugged in and tested.

10. Matomo is up.

The biggest issue we had was that networking on the servers didn't work properly because they were booted before the router was configured properly. We eventually got the router working properly and expected the servers to immediately come online. They didn't. I'm sure there was a command we could run to reset the networks, but the easier thing was to simply reboot, so we did that.

All in all, it was fairly smooth. No *major* issues and things like networking worked well. Could have been smoother, but not bad for a once-in-a-decade event.

**Comment by mlissner** [comment]
- [x] The last piece to check is that our replication client computers are still accessible from the new IP. I asked them to set this up awhile ago, but I don't know if they have. 

They have!
